signac/synced_collections/backends/collection_json.py

```diff
@@ -0,0 +1,387 @@
+# TODO: This method should be removed in signac 2.0.
+def _convert_key_to_str(data):
+    """Recursively convert non-string keys to strings in dicts.
+
+    This method supports :class:`collections.abc.Sequence` or
+    :class:`collections.abc.Mapping` types as inputs, and recursively
+    searches for any entries in :class:`collections.abc.Mapping` types where
+    the key is not a string. This functionality is added for backwards
+    compatibility with legacy behavior in signac, which allowed integer keys
+    for dicts. These inputs were silently converted to string keys and stored
+    since JSON does not support integer keys. This behavior is deprecated and
+    will become an error in signac 2.0.
+
+    Note for developers: this method is designed for use as a validator in the
+    synced collections framework, but due to the backwards compatibility requirement
+    it violates the general behavior of validators by modifying the data in place.
+    This behavior can be removed in signac 2.0 once non-str keys become an error.
+    """
+    if isinstance(data, dict):
+
+        def _str_key(key):
+            if not isinstance(key, str):
+                warnings.warn(
+                    f"Use of {type(key).__name__} as key is deprecated "
+                    "and will be removed in version 2.0",
+                    DeprecationWarning,
+                )
+                key = str(key)
+            return key
+
+        # Get a list of keys a priori to support modification in place.
+        for key in list(data):
+            _convert_key_to_str(data[key])
+            data[_str_key(key)] = data.pop(key)
+    elif isinstance(data, list):
+        for i, value in enumerate(data):
+            _convert_key_to_str(value)
+
+
+class JSONCollection(SyncedCollection):
+    """A :class:`~.SyncedCollection` that synchronizes with a JSON file.
+
+    This collection implements synchronization by reading and writing the associated
+    JSON file in its entirety for every read/write operation. This backend is a good
+    choice for maximum accessibility and transparency since all data is immediately
+    accessible in the form of a text file with no additional tooling, but is
+    likely a poor choice for high performance applications.
+
+    **Thread safety**
+
+    The :class:`JSONCollection` is thread-safe. To make these collections safe, the
+    ``write_concern`` flag is ignored in multithreaded execution, and the
+    write is **always** performed via a write to temporary file followed by a
+    replacement of the original file. The file replacement operation uses
+    :func:`os.replace`, which is guaranteed to be atomic by the Python standard.
+
+    Parameters
+    ----------
+    filename : str
+        The filename of the associated JSON file on disk.
+    write_concern : bool, optional
+        Ensure file consistency by writing changes back to a temporary file
+        first, before replacing the original file (Default value = False).
```

Comment:
The reason to include `*args` and `**kwargs` here is that it's the only way to properly handle the arbitrary multiple inheritance patterns that synced collections are built on, because all arguments are passed to all constructors by super. I don't think documenting these makes sense, because a user should only see the docstrings for arguments that a given class knows how to parse (based on its inheritance hierarchy). For example, `JSONDict` documents `filename` and `write_concern` (handled by `JSONCollection`), `parent` (handled by `SyncedCollection`), and `data` (handled by `SyncedDict`).

Reply:
The behavior of what happens with args and kwargs is not obvious. In most classes like this, I write explicit docstrings that say args and kwargs are forwarded to the parent constructor.

Reply:
That's fine, but it will need to be copy-pasted to every single subclass of `SyncedCollection`.